cpp/kiss_icp/core/Preprocessing.cpp

```diff
@@ -28,42 +28,22 @@
 #include <Eigen/Core>
 #include <algorithm>
 #include <cmath>
+#include <kiss_icp/core/VoxelUtils.hpp>
 #include <sophus/se3.hpp>
 #include <vector>
 
-namespace {
-// TODO(all): Maybe try to merge these voxel uitls with VoxelHashMap implementation
-using Voxel = Eigen::Vector3i;
-struct VoxelHash {
-    size_t operator()(const Voxel &voxel) const {
-        const uint32_t *vec = reinterpret_cast<const uint32_t *>(voxel.data());
-        return ((1 << 20) - 1) & (vec[0] * 73856093 ^ vec[1] * 19349669 ^ vec[2] * 83492791);
-    }
-};
-
-Voxel PointToVoxel(const Eigen::Vector3d &point, double voxel_size) {
-    return Voxel(static_cast<int>(std::floor(point.x() / voxel_size)),
-                 static_cast<int>(std::floor(point.y() / voxel_size)),
-                 static_cast<int>(std::floor(point.z() / voxel_size)));
-}
-}  // namespace
-
 namespace kiss_icp {
 std::vector<Eigen::Vector3d> VoxelDownsample(const std::vector<Eigen::Vector3d> &frame,
                                              double voxel_size) {
-    tsl::robin_map<Voxel, Eigen::Vector3d, VoxelHash> grid;
+    tsl::robin_map<Voxel, Eigen::Vector3d> grid;
     grid.reserve(frame.size());
-    for (const auto &point : frame) {
+    std::for_each(frame.cbegin(), frame.cend(), [&](const auto &point) {
         const auto voxel = PointToVoxel(point, voxel_size);
-        if (grid.contains(voxel)) continue;
-        grid.insert({voxel, point});
-    }
-    std::vector<Eigen::Vector3d> frame_dowsampled;
-    frame_dowsampled.reserve(grid.size());
-    for (const auto &[voxel, point] : grid) {
-        (void)voxel;
-        frame_dowsampled.emplace_back(point);
-    }
+        if (!grid.contains(voxel)) grid.insert({voxel, point});
+    });
+    std::vector<Eigen::Vector3d> frame_dowsampled(grid.size());
```

Comment:
This is not the same as before, and I believe it's slower. This will call the default constructor of `Eigen::Vector3d`, `grid.size()` times, whereas before, we just reserved memory.

Additionally, this is not a transformation operation, and it reads super confusing to me.

That being said if we decide to stick to `std::transform`, then we should use constant iterators `cbegin`, `cend`

EDIT: It's slower, see Quickbench [here](https://quick-bench.com/q/Y7ZmTonpYfD7EpY8iYNYHoqWM4I)
![image](https://github.com/PRBonn/kiss-icp/assets/21349875/699e55d3-7ab3-498e-ad8a-dd7ed855e317)

Reply:
I had a (potentially) better idea using move semantics; I can show you the results here.

![image](https://github.com/PRBonn/kiss-icp/assets/37455909/8bba9332-89b9-4f03-ac45-ea5b25c0e636)


This is the corresponding [Quickbench](https://quick-bench.com/q/Bx2WNJkfPctQkFykvob0nKRrrAk)

Reply:
And indeed, you are seeing a faster execution because you removed the `static` from the test data generation, if you put it back you will see a slower execution on all this move things

Reply:
Why it was ```static``` in the first place?

Reply:
Mainly because the results change a lot based on that I am not sure which one is more efficient at this point; maybe the ```std::make_move_iterator```is too much for new users, but also the cast to void in the middle of the loop it is not much better, as it is hugely as fuck, if you have a strong opinion for the ```emplace_back``` instead of the new version (I consider the benchmark results inconclusive, as in reality we don't use a ```static``` data structure, I will rather trust my version of the benchmark) I will make it at least a ```for_each```

Reply:
For me, any of the two is fine; the current version is just ugly period

Reply:
So, the static was a quick way to share the data among benchmarks that of course didn't work. So forget about quick bench for this situation. I wrote myself a new benchmark and indeed the results points to the direction that the execution is all the same,  so it's a question of readability, not performance.
![image](https://github.com/PRBonn/kiss-icp/assets/21349875/a0141095-e397-42d9-9cf9-c721d32baf1e)

```cpp
#include <benchmark/benchmark.h>

#include <Eigen/Core>
#include <algorithm>
#include <iostream>
#include <memory>
#include <random>
#include <vector>

// Global shared data
std::shared_ptr<const std::vector<Eigen::Vector3d>> shared_data = nullptr;

void InitializeSharedData() {
    if (!shared_data) {
        std::cerr << ">>>>>>>>>>>>>>>>> Creating Shared Data <<<<<<<<<<<<<<<<<" << std::endl;
        std::vector<Eigen::Vector3d> data;

        // Random number generation
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<> dis(-100.0, 100.0);

        for (int i = 0; i < 1 << 20; ++i) {
            double x = dis(gen);
            double y = dis(gen);
            double z = dis(gen);
            data.push_back({x, y, z});
        }

        shared_data = std::make_shared<const std::vector<Eigen::Vector3d>>(std::move(data));
    }
}

static void ReserveEmplaceBack(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> grid = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled;
        frame_dowsampled.reserve(grid.size());
        for (const auto point : grid) {
            frame_dowsampled.emplace_back(point);
        }
        benchmark::DoNotOptimize(frame_dowsampled);
    }
}
static void WithMove(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> grid = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled(grid.size());
        std::transform(grid.begin(), grid.end(), frame_dowsampled.begin(),
                       [](auto &point) { return point; });
        benchmark::DoNotOptimize(frame_dowsampled);
    }
}

static void StdCopy(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> grid = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled(grid.size());
        std::copy(grid.cbegin(), grid.cend(), frame_dowsampled.begin());
        benchmark::DoNotOptimize(frame_dowsampled);
    }
}

static void StdCopyMove(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> grid = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled(grid.size());
        std::copy(std::make_move_iterator(grid.cbegin()), std::make_move_iterator(grid.cend()),
                  frame_dowsampled.begin());
        benchmark::DoNotOptimize(frame_dowsampled);
    }
}

static void PlainMove(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> grid = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled(grid.size());
        std::move(grid.cbegin(), grid.cend(), frame_dowsampled.begin());
        benchmark::DoNotOptimize(frame_dowsampled);
    }
}

BENCHMARK(ReserveEmplaceBack)->Unit(benchmark::kMillisecond);
BENCHMARK(WithMove)->Unit(benchmark::kMillisecond);
BENCHMARK(StdCopy)->Unit(benchmark::kMillisecond);
BENCHMARK(StdCopyMove)->Unit(benchmark::kMillisecond);
BENCHMARK(PlainMove)->Unit(benchmark::kMillisecond);

int main(int argc, char **argv) {
    InitializeSharedData();
    ::benchmark::Initialize(&argc, argv);
    ::benchmark::RunSpecifiedBenchmarks();
}
```

Reply:
Actually, as always, STL for the win on everything 

![image](https://github.com/PRBonn/kiss-icp/assets/37455909/1c30333e-79c2-4f18-95e1-f670d4b1541e)

The improvement I saw is due to STL algorithm optimization rather than the move logic itself. 

As always thanks to all the gods for STL.

Reply:
For reference [this](https://quick-bench.com/q/PKnGsB9xDjYHn8L95_O_2anDI_o) is the benchmark

Reply:
Adding ```static``` again...

![image](https://github.com/PRBonn/kiss-icp/assets/37455909/031475b7-ced7-42cd-9840-f3d01901398e)

Reply:
Actually not, I suggest we stop using this quick bench garbage as it gives contracictiorady results. I tested this locally, and the runtime is literally the same.

![image](https://github.com/PRBonn/kiss-icp/assets/21349875/787338fc-647a-41a8-bd1c-6738910d74cd)

```cpp
#include <benchmark/benchmark.h>
#include <tsl/robin_map.h>

#include <Eigen/Core>
#include <algorithm>
#include <iostream>
#include <memory>
#include <random>
#include <vector>

#include "VoxelUtils.hpp"
using namespace kiss_icp;

// Global shared data
std::shared_ptr<const std::vector<Eigen::Vector3d>> shared_data = nullptr;

void InitializeSharedData() {
    if (!shared_data) {
        std::cerr << ">>>>>>>>>>>>>>>>> Creating Shared Data <<<<<<<<<<<<<<<<<" << std::endl;
        std::vector<Eigen::Vector3d> data;

        // Random number generation
        std::random_device rd;
        std::mt19937 gen(rd());
        std::uniform_real_distribution<> dis(-100.0, 100.0);

        for (int i = 0; i < 1 << 10; ++i) {
            double x = dis(gen);
            double y = dis(gen);
            double z = dis(gen);
            data.push_back({x, y, z});
        }

        shared_data = std::make_shared<const std::vector<Eigen::Vector3d>>(std::move(data));
    }
}

std::vector<Eigen::Vector3d> VoxelDownsampleNew(const std::vector<Eigen::Vector3d> &frame,
                                                double voxel_size) {
    tsl::robin_map<Voxel, Eigen::Vector3d> grid;
    grid.reserve(frame.size());
    std::for_each(frame.cbegin(), frame.cend(), [&](const auto &point) {
        const auto voxel = PointToVoxel(point, voxel_size);
        if (!grid.contains(voxel)) grid.insert({voxel, point});
    });

    std::vector<Eigen::Vector3d> frame_dowsampled;
    frame_dowsampled.reserve(grid.size());
    for (const auto &voxel_and_point : grid) frame_dowsampled.emplace_back(voxel_and_point.second);

    return frame_dowsampled;
}

std::vector<Eigen::Vector3d> VoxelDownsampleOld(const std::vector<Eigen::Vector3d> &frame,
                                                double voxel_size) {
    tsl::robin_map<Voxel, Eigen::Vector3d> grid;
    grid.reserve(frame.size());
    std::for_each(frame.cbegin(), frame.cend(), [&](const auto &point) {
        const auto voxel = PointToVoxel(point, voxel_size);
        if (!grid.contains(voxel)) grid.insert({voxel, point});
    });

    std::vector<Eigen::Vector3d> frame_dowsampled(grid.size());
    std::transform(grid.cbegin(), grid.cend(), frame_dowsampled.begin(),
                   [](const auto &voxel_and_point) { return voxel_and_point.second; });

    return frame_dowsampled;
}

std::vector<Eigen::Vector3d> VoxelDownsampleForEach(const std::vector<Eigen::Vector3d> &frame,
                                                    double voxel_size) {
    tsl::robin_map<Voxel, Eigen::Vector3d> grid;
    grid.reserve(frame.size());
    std::for_each(frame.cbegin(), frame.cend(), [&](const auto &point) {
        const auto voxel = PointToVoxel(point, voxel_size);
        if (!grid.contains(voxel)) grid.insert({voxel, point});
    });

    std::vector<Eigen::Vector3d> frame_dowsampled;
    frame_dowsampled.reserve(grid.size());
    std::for_each(grid.cbegin(), grid.cend(), [&](const auto &voxel_and_point) {
        frame_dowsampled.emplace_back(voxel_and_point.second);
    });

    return frame_dowsampled;
}

static void NewVoxelDownsample(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> frame = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled = VoxelDownsampleNew(frame, 1.0);
    }
}

static void OldVoxelDownsample(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> frame = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled = VoxelDownsampleOld(frame, 1.0);
    }
}
static void ForEach(benchmark::State &state) {
    for (auto _ : state) {
        std::vector<Eigen::Vector3d> frame = *shared_data;
        std::vector<Eigen::Vector3d> frame_dowsampled = VoxelDownsampleForEach(frame, 1.0);
    }
}

BENCHMARK(NewVoxelDownsample);
BENCHMARK(OldVoxelDownsample);
BENCHMARK(ForEach);

int main(int argc, char **argv) {
    InitializeSharedData();
    ::benchmark::Initialize(&argc, argv);
    ::benchmark::RunSpecifiedBenchmarks();
}
```

Reply:
Make sense now, to confirm, I was testing on real data:

### **main**

![image](https://github.com/PRBonn/kiss-icp/assets/37455909/2c62e4d9-6832-41f4-8d80-82b513ee9dce)

### **transform**

![image](https://github.com/PRBonn/kiss-icp/assets/37455909/909e7287-8052-4850-93c9-3b1d4b0cef00)

### **for_each**

![image](https://github.com/PRBonn/kiss-icp/assets/37455909/27ab7aac-37de-401c-b92b-a3856cbb19e1)

Reply:
Literally same, shall we go for the ```for_each```?

Reply:
Yes!